app/models/message.rb

```diff
@@ -45,6 +45,16 @@ def not_finished?
     !finished?
   end
 
+  def input_token_count=(count)
+    self[:input_token_cost] = assistant.language_model.input_token_cost_cents * count
+    super(count)
+  end
+
+  def output_token_count=(count)
+    self[:output_token_cost] = assistant.language_model.output_token_cost_cents * count
+    super(count)
+  end
+
```

Comment:
I see what you're doing here, but the more idiomatic way is:

```
before_save :check_age_change, if: :age_changed?

def check_age_change
  # Your code here
end
```

You could do it as two separate callbacks or as one, e.g.:

`before_save :update_token_costs, if: ->{ input_token_cost_changed? || output_token_cost_changed? }`

I'd probably do it as two separate callbacks with two separate helper methods:

```
before_save :update_output_token_cost, if: :output_token_cost_changed?
before_save :update_input_token_cost, if: :input_token_cost_changed?
```

It's probably a little more dangerous to override the setter like this (i.e. there may be some unintended side effects to general activerecord behavior). But my main motivation for doing it is just because it's what other rails devs would expect to see. You're used to looking for hooks at the top of a model to familiarize your self with any surprise "side effects", whereas it's a bit unexpected for the setter to have a ripple like this.

Reply:
Let's add a couple unit test for this too. I think we want to prove that setting either of these counts properly updates the cost *and* that setting one of these without setting the other does not zero out the unchanging one.